src/test/lib/DataTest.cpp

```diff
@@ -1857,3 +1857,302 @@ QuicTestReceiveResumeNoData(
         }
     }
 }
+
+struct AckSendDelayTestContext {
+    AckSendDelayTestContext() :
+        SendBuffer(1, 200)
+    {};
+    HQUIC ServerConfiguration;
+    QuicSendBuffer SendBuffer;
+    EventScope ServerStreamStartedEvent;
+    EventScope ClientReceiveDataEvent;
+    EventScope ClientConnectedEvent;
+    ConnectionScope ServerConnection;
+    ConnectionScope ClientConnection;
+    StreamScope ServerStream;
+    StreamScope ClientStream;
+    uint64_t AckCountStart;
+    uint64_t AckCountStop;
+};
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+_Function_class_(QUIC_STREAM_CALLBACK)
+static
+QUIC_STATUS
+QUIC_API
+QuicAckDelayStreamHandler(
+    _In_ HQUIC QuicStream,
+    _In_opt_ void* Context,
+    _Inout_ QUIC_STREAM_EVENT* Event
+    )
+{
+    QUIC_STATUS Status = QUIC_STATUS_SUCCESS;
+    AckSendDelayTestContext* TestContext = (AckSendDelayTestContext*)Context;
+    if (TestContext->ServerStream.Handle == QuicStream) {
+        //
+        // Server side
+        //
+        switch (Event->Type) {
+        case QUIC_STREAM_EVENT_RECEIVE:
+            Event->RECEIVE.TotalBufferLength = 0;
+            Status = MsQuic->StreamSend(
+                QuicStream,
+                TestContext->SendBuffer.Buffers,
+                TestContext->SendBuffer.BufferCount,
+                QUIC_SEND_FLAG_FIN,
+                nullptr);
+            if (QUIC_FAILED(Status)) {
+                TEST_FAILURE("Server failed to send to send data back 0x%x", Status);
```

Comment:
`TEST_FAILURE` doesn't return. You should put explicit return calls after each of them in this function.